internal/pkg/term/syncbuffer/termprinter.go

```diff
@@ -0,0 +1,184 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package syncbuffer provides a goroutine safe bytes.Buffer as well printing functionality to the terminal.
+package syncbuffer
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/aws/copilot-cli/internal/pkg/term/cursor"
+	"golang.org/x/term"
+)
+
+// pollIntervalForSyncBuffer is the time Interval to wait between checking whether the output buffers are done.
+const pollIntervalForSyncBuffer = 60 * time.Millisecond
+
+// FileWriter is the interface to write to a file.
+type FileWriter interface {
+	io.Writer
+}
+
+// LabeledTermPrinter is a printer to display label and logs to the terminal.
+type LabeledTermPrinter struct {
+	term             FileWriter           // term writes logs to the terminal FileWriter.
+	buffers          []*LabeledSyncBuffer // buffers stores logs before writing to the terminal.
+	numLines         int                  // number of lines that has to be written from each buffer.
+	padding          int                  // Leading spaces before rendering to terminal.
+	prevWrittenLines int                  // number of lines written from all the buffers.
+	termWidth        int                  // width of terminal.
+}
+
+// LabeledTermPrinterOption is a type alias to configure LabeledTermPrinter.
+type LabeledTermPrinterOption func(ltp *LabeledTermPrinter)
+
+// NewLabeledTermPrinter returns a LabeledTermPrinter that can print to the terminal filewriter from buffers.
+func NewLabeledTermPrinter(fw FileWriter, bufs []*LabeledSyncBuffer, opts ...LabeledTermPrinterOption) (*LabeledTermPrinter, error) {
+	width, err := terminalWidth()
+	if err != nil {
+		return nil, fmt.Errorf("get terminal width: %w", err)
+	}
+
+	ltp := &LabeledTermPrinter{
+		term:      fw,
+		buffers:   bufs,
+		termWidth: width,
+		numLines:  -1, // By default set numlines to -1 to print all from buffers.
+	}
+	for _, opt := range opts {
+		opt(ltp)
+	}
+	return ltp, nil
+}
+
+// WithNumLines sets the numlines of LabeledTermPrinter.
+func WithNumLines(n int) LabeledTermPrinterOption {
+	return func(ltp *LabeledTermPrinter) {
+		ltp.numLines = n
+	}
+}
+
+// WithPadding sets the padding of LabeledTermPrinter.
+func WithPadding(n int) LabeledTermPrinterOption {
+	return func(ltp *LabeledTermPrinter) {
+		ltp.padding = n
+	}
+}
+
+// isDone returns true if all the buffers are done else returns false.
+func (ltp *LabeledTermPrinter) isDone() bool {
+	for _, buf := range ltp.buffers {
+		if !buf.syncBuf.IsDone() {
+			return false
+		}
+	}
+	return true
+}
+
+// Print prints the label and the last N lines of logs from each buffer to the termPrinter fileWriter.
+// If numLines is -1 then print all the values from buffers.
+// It polls each buffer until all the buffers are marked done,
+// and erases the previous output after sleeping for a short duration.
+func (ltp *LabeledTermPrinter) Print() {
+	if ltp.numLines == -1 {
```

Comment:
nit: consider moving [magic numbers](https://refactoring.com/catalog/replaceMagicLiteral.html) to consts:

```go
const (
   printAllLinesInBuf = -1
)

```